Core/src/Geometry/Extent.cpp

```diff
@@ -1,27 +1,159 @@
 // This file is part of the Acts project.
 //
-// Copyright (C) 2020 CERN for the benefit of the Acts project
+// Copyright (C) 2022 CERN for the benefit of the Acts project
 //
 // This Source Code Form is subject to the terms of the Mozilla Public
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
 #include "Acts/Geometry/Extent.hpp"
 
-#include <iomanip>
-#include <ostream>
+#include "Acts/Utilities/Helpers.hpp"
+
+Acts::Extent::Extent(
+    const std::array<std::array<ActsScalar, 2>, binValues>& envelope)
+    : m_constrains(0), m_envelope(envelope) {
+  m_range[binR] =
+      Range1D<ActsScalar>(0., std::numeric_limits<ActsScalar>::max());
+  m_range[binPhi] = Range1D<ActsScalar>(-M_PI, M_PI);
+  m_range[binRPhi] =
+      Range1D<ActsScalar>(0., std::numeric_limits<ActsScalar>::max());
+  m_range[binMag] =
+      Range1D<ActsScalar>(0., std::numeric_limits<ActsScalar>::max());
+}
+
+void Acts::Extent::extend(const Vector3& vtx,
+                          const std::vector<BinningValue>& bValues,
+                          bool applyEnv, bool fillHistograms) {
+  auto adapt = [&](BinningValue bValue) {
+    // Get the casted value given the binnin value description
+    ActsScalar cValue = VectorHelpers::cast(vtx, bValue);
+    if (fillHistograms) {
+      m_valueHistograms[bValue].push_back(cValue);
+    }
+    // Apply envelope as suggested
+    ActsScalar lEnv = applyEnv ? m_envelope[bValue][0] : 0.;
+    ActsScalar hEnv = applyEnv ? m_envelope[bValue][1] : 0.;
+    ActsScalar mValue = cValue - lEnv;
+    // Special protection for radial value
+    if (bValue == binR and mValue < 0.) {
+      mValue = std::max(mValue, 0.);
+    }
+    if (constrains(bValue)) {
+      m_range[bValue].expand(mValue, cValue + hEnv);
+    } else {
+      m_range[bValue].shrink(mValue, cValue + hEnv);
+    }
+    m_constrains.set(bValue);
+  };
+
+  for (auto bValue : bValues) {
+    adapt(bValue);
```

Comment:
Is there a reason for this to be an anonymous function rather than just putting the code in the for loop directly?

Reply:
Not really, will change it.